### PR TITLE
Migrating data to datastore

### PR DESCRIPTION
**Pull request by maxsam4**
- Optimized Datastore (From 23.5kb to 11.5kb with 4 extra functions)
- Moved investor list to datastore
- Moved canBuy logic from GTM to STO
- Added Investor flags
- Renamed whitelist data to kyc data and moved flags to separate a uint256
- Added eth gas reporter (it reports gas usage in tests, run using `npm run gas`)

### CODE REVIEW

**Comment by F-OBrien** [comment]
I had a read through this pull request and wanted to reference a feature request I had raised #532 
related to canBuyFromSTO. 
This pull request goes part of the way implementing my suggestion by removing the canbBuyFromSTO check from the GTM and including it in the STO contracts (and also moving the accredited status out of the STO contract) but I believe it is still the case that setting canBuy for an investor allows them to buy from all STO's and does not allow the possibility to apply different investor restrictions to different STO's. 
I still think the ability to have concurrent offerings with certain investors only allowed to invest in one or the other may be a useful feature in a future pull request if not in this one.

**Comment by maxsam4** [comment]
Thanks for your feedback @F-OBrien! Different restrictions on different simultaneous STOs is a planned feature.

File: contracts/libraries/VersionUtils.sol [hunk 1]
```diff
@@ -110,21 +110,18 @@ library VersionUtils {
     /**
      * @notice Used to packed the KYC data
      */
-    function packKYC(uint64 _a, uint64 _b, uint64 _c, uint8 _d, uint8 _e, uint8 _f) internal pure returns(uint256) {
-        return (uint256(_a) << 152) | (uint256(_b) << 88) | (uint256(_c) << 24) | (uint256(_d) << 16) | (uint256(_e) << 8) | uint256(_f);
+    function packKYC(uint64 _a, uint64 _b, uint64 _c, uint8 _d) internal pure returns(uint256) {
```
Comment (adamdossa) [inline contracts/libraries/VersionUtils.sol:113]: Could we have a comment explaining the function here a bit for auditors / people reviewing the code etc.

File: contracts/modules/STO/USDTieredSTO.sol [hunk 2]
```diff
@@ -568,6 +544,17 @@ contract USDTieredSTO is USDTieredSTOStorage, STO {
...
     }
 
+    function _isAccredited(address _investor) internal view returns(bool) {
+        IDataStore dataStore = IDataStore(getDataStore());
+        return _getIsAccredited(_investor, dataStore);
+    }
+
+    function _getIsAccredited(address _investor, IDataStore dataStore) internal view returns(bool) {
+        uint256 flags = dataStore.getUint256(_getKey(INVESTORFLAGS, _investor));
+        uint256 flag = flags & uint256(1); //isAccredited is flag 0 so we don't need to bit shift flags.
```
Comment (adamdossa) [inline contracts/modules/STO/USDTieredSTO.sol:554]: Should we store the meaning for each bit in INVESTORFLAGS somewhere? Even if it is just in a comment in the DataStore, or a .md file somewhere in the repo.

**Review by adamdossa — APPROVED** [review]

File: contracts/modules/STO/USDTieredSTO.sol (hunk 2 above)
Comment (maxsam4) [reply contracts/modules/STO/USDTieredSTO.sol:554]: I already added it https://github.com/PolymathNetwork/polymath-core/blob/cdc941da0b7851f965a396b87c0dc212d5ec8296/docs/investor_flags.md :)

File: contracts/modules/STO/STOStorage.sol [hunk 3]
```diff
@@ -5,6 +5,7 @@ pragma solidity ^0.5.0;
  */
 
 contract STOStorage {
+    bytes32 internal constant INVESTORFLAGS = "INVESTORFLAGS";
```
Comment (satyamakgec) [inline contracts/modules/STO/STOStorage.sol:8]: Here we can also add the hex of the string

File: contracts/modules/STO/USDTieredSTO.sol (hunk 2 above)
Comment (satyamakgec) [reply contracts/modules/STO/USDTieredSTO.sol:554]: @maxsam4 a minor typo fix in the .md file

File: contracts/modules/STO/STOStorage.sol (hunk 3 above)
Comment (maxsam4) [reply contracts/modules/STO/STOStorage.sol:8]: The key is not sha3("INVESTORFLAGS") but sha3("INVESTORFLAGS" + $Investor_address) therefore, we can't pre-compute it.

File: contracts/modules/STO/USDTieredSTO.sol (hunk 2 above)
Comment (maxsam4) [reply contracts/modules/STO/USDTieredSTO.sol:554]: Pushed a fix for the typo I found. If there is any other typo present, please let me know.

**Review by satyamakgec — APPROVED** [review]

File: contracts/datastore/DataStore.sol [hunk 4]
```diff
@@ -15,22 +15,26 @@ contract DataStore is DataStoreStorage, IDataStore {
...
     modifier onlyAuthorized() {
-        bool isOwner = msg.sender == IOwnable(address(securityToken)).owner();
-        require(isOwner ||
-            securityToken.isModule(msg.sender, DATA_KEY) ||
-            securityToken.checkPermission(msg.sender, address(this), MANAGEDATA),
+        _isAuthorized();
+        _;
+    }
+
+    function _isAuthorized() internal view {
```
Comment (Ttokkiyo) [inline contracts/datastore/DataStore.sol:17]: ok lov u

File: contracts/storage/USDTieredSTOStorage.sol [hunk 5]
```diff
@@ -30,7 +30,7 @@ contract USDTieredSTOStorage {
         uint256 mintedDiscountPoly;
     }
 
-    mapping(address => uint256) nonAccreditedLimitUSDOverride;
+    mapping(address => uint256) public nonAccreditedLimitUSDOverride;
```
Comment (Ttokkiyo) [inline contracts/storage/USDTieredSTOStorage.sol:33]: The mapping address is wrong D:

Comment (satyamakgec) [reply contracts/storage/USDTieredSTOStorage.sol:33]: What do you mean ?

File: contracts/datastore/DataStoreStorage.sol [hunk 6]
```diff
@@ -16,6 +16,6 @@ contract DataStoreStorage {
     mapping (bytes32 => address[]) internal addressArrayData;
     mapping (bytes32 => bool[]) internal boolArrayData;
 
-    uint8 constant DATA_KEY = 6;
-    bytes32 public constant MANAGEDATA = "MANAGEDATA";
+    uint8 internal constant DATA_KEY = 6;
+    bytes32 internal constant MANAGEDATA = "MANAGEDATA";
```
Comment (akashvekariya) [inline contracts/datastore/DataStoreStorage.sol:20]: then this two variables won't be accessible

**Pull request merged**